File: python/binary_tree_max_path_sum.py
"""
Definition of TreeNode:
""" 
import sys 
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        # write your code here
        res, tmp = self.getMax(root)
        
        return res
        
    def getMax(self, root):
        if root is None:
            return -sys.maxsize-1, 0
        left = self.getMax(root.left)
        right = self.getMax(root.right)
        res = max(left[0], right[0], left[1]+right[1]+root.val)
        tmp = max(left[1]+root.val, right[1]+root.val, 0)
        
        return res,tmp

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)

test=Solution()
res = test.maxPathSum(root)
print(res)

File: cpp/favourite/BinaryTreeMaxPathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
   
public:
    int maxPathSum(TreeNode* root) {
        if(!root) return 0;
        int res =INT_MIN;
        getMax(root, res);
        // maxPathSum(root->left);
        // maxPathSum(root->right);
        return res;
        
    }
    
    int getMax(TreeNode* root, int& res) {
        if(!root) return 0;
       
        int left = max(getMax(root->left, res),0);
        int right = max(getMax(root->right, res), 0);
       
        res = max(res, root->val+left+right);
        return root->val+max(left, right);
        
    }
    
};